File: src/search_provider.py
# Copyright (C) 2022 - 2025 Alessandro Iepure
#
# SPDX-License-Identifier: GPL-3.0-or-later

from gi.repository import Gio, GLib, GObject
import subprocess
import time
from typing import List, Dict, Any
from .tools import TOOLS_METADATA, search_tools


class DevToolboxSearchProvider(GObject.Object):
    """GNOME Shell search provider for Dev Toolbox."""

    IDLE_TIMEOUT = 10  # 30 seconds of inactivity
    CACHE_DURATION = 300  # 5 minutes

    def __init__(self):
        super().__init__()
        self._search_cache = {}
        self._cache_timestamps = {}
        self._bus = None
        self._registration_id = None
        self._idle_timeout_id = None
        self._main_loop = None

        # D-Bus interface
        self.interface_xml = """
        <node>
            <interface name="org.gnome.Shell.SearchProvider2">
                <method name="GetInitialResultSet">
                    <arg type="as" name="terms" direction="in" />
                    <arg type="as" name="results" direction="out" />
                </method>
                <method name="GetSubsearchResultSet">
                    <arg type="as" name="previous_results" direction="in" />
                    <arg type="as" name="terms" direction="in" />
                    <arg type="as" name="results" direction="out" />
                </method>
                <method name="GetResultMetas">
                    <arg type="as" name="identifiers" direction="in" />
                    <arg type="aa{sv}" name="metas" direction="out" />
                </method>
                <method name="ActivateResult">
                    <arg type="s" name="identifier" direction="in" />
                    <arg type="as" name="terms" direction="in" />
                    <arg type="u" name="timestamp" direction="in" />
                </method>
                <method name="LaunchSearch">
                    <arg type="as" name="terms" direction="in" />
                    <arg type="u" name="timestamp" direction="in" />
                </method>
            </interface>
        </node>
        """

    def _reset_idle_timer(self):
        """Reset the idle timeout timer"""
        
        # Cancel existing timer
        if self._idle_timeout_id:
            GLib.source_remove(self._idle_timeout_id)
        
        # Start new timer
        self._idle_timeout_id = GLib.timeout_add_seconds(
            self.IDLE_TIMEOUT, 
            self._on_idle_timeout
        )
        
        print(f"DEBUG: Reset idle timer to {self.IDLE_TIMEOUT} seconds")

    def _on_idle_timeout(self):
        """Called when idle timeout expires"""
        
        print("DEBUG: Idle timeout reached, shutting down search provider...")
        if self._main_loop:
            self._main_loop.quit()
        return False  # No repeat

    def _search_with_cache(self, terms: List[str]) -> List[str]:
        """Search with caching for performance"""

        cache_key = '|'.join(sorted(terms))
        current_time = time.time()

        # Check cache
        if (cache_key in self._search_cache and
            cache_key in self._cache_timestamps and
                current_time - self._cache_timestamps[cache_key] < self.CACHE_DURATION):
            print(f"DEBUG: Cache hit for: {cache_key}")
            return self._search_cache[cache_key]

        # Perform search
        results = search_tools(terms)[:8]  # Limit to 8 results for GNOME Shell
        print(f"DEBUG: Search results for '{' '.join(terms)}': {results}")

        # Cache results
        self._search_cache[cache_key] = results
        self._cache_timestamps[cache_key] = current_time

        # Cleanup old cache entries
        self._cleanup_cache(current_time)

        return results

    def _cleanup_cache(self, current_time: float):
        """Remove old cache entries"""

        expired_keys = [
            key for key, timestamp in self._cache_timestamps.items()
            if current_time - timestamp >= self.CACHE_DURATION
        ]
        for key in expired_keys:
            self._search_cache.pop(key, None)
            self._cache_timestamps.pop(key, None)

    def GetInitialResultSet(self, terms: List[str]) -> List[str]:
        """GNOME Shell SearchProvider2 interface method"""

        print(f"DEBUG: GetInitialResultSet called with: {terms}")
        self._reset_idle_timer()
        return self._search_with_cache(terms)

    def GetSubsearchResultSet(self, previous_results: List[str], terms: List[str]) -> List[str]:
        """GNOME Shell SearchProvider2 interface method"""

        print(f"DEBUG: GetSubsearchResultSet called with: {terms}")
        self._reset_idle_timer()
        return self._search_with_cache(terms)

    def GetResultMetas(self, identifiers: List[str]) -> List[Dict[str, Any]]:
        """GNOME Shell SearchProvider2 interface method"""

        print(f"DEBUG: GetResultMetas called with: {identifiers}")
        self._reset_idle_timer()
        
        metas = []
        for tool_id in identifiers:
            if tool_id in TOOLS_METADATA:
                tool = TOOLS_METADATA[tool_id]

                meta = {
                    'id': GLib.Variant('s', tool_id),
                    'name': GLib.Variant('s', tool['title']),
                    'description': GLib.Variant('s', tool['tooltip']),
                }
                metas.append(meta)
                print(f"DEBUG: Added meta for {tool_id}: {tool['title']}")
        return metas

    def ActivateResult(self, identifier: str, terms: List[str], timestamp: int):
        """GNOME Shell SearchProvider2 interface method"""

        print(
            f"DEBUG: ActivateResult called - tool: {identifier}, terms: {terms}")
        self._reset_idle_timer()
        
        # Launch app with specific tool
        try:
            subprocess.run(
                [
                    '/app/bin/devtoolbox',
                    '--tool', identifier
                ],
                check=True,
            )

        except Exception as e:
            print(f"ERROR: Failed to launch app: {e}")

    def LaunchSearch(self, terms: List[str], timestamp: int):
        """GNOME Shell SearchProvider2 interface method"""

        print(f"DEBUG: LaunchSearch called with: {terms}")
        self._reset_idle_timer()
        
        try:
            subprocess.run(
                [
                    '/app/bin/devtoolbox',
                    '--search', ' '.join(terms)
                ],
                check=True,
            )
        except Exception as e:
            print(f"ERROR: Failed to launch search: {e}")

    def run(self):
        """Start the search provider D-Bus service"""

        print("DEBUG: Starting search provider D-Bus service...")

        try:
            # Get the session bus
            self._bus = Gio.bus_get_sync(Gio.BusType.SESSION, None)
            print("DEBUG: Connected to session bus")

            # Register the interface
            node_info = Gio.DBusNodeInfo.new_for_xml(self.interface_xml)
            interface_info = node_info.interfaces[0]
            print("DEBUG: Parsed interface XML")

            def handle_method_call(connection, sender, object_path, interface_name, method_name, parameters, invocation):
                print(f"DEBUG: Method called: {method_name}")
                try:
                    if method_name == "GetInitialResultSet":
                        terms = parameters.unpack()[0]
                        results = self.GetInitialResultSet(terms)
                        invocation.return_value(
                            GLib.Variant('(as)', (results,)))

                    elif method_name == "GetSubsearchResultSet":
                        previous_results, terms = parameters.unpack()
                        results = self.GetSubsearchResultSet(
                            previous_results, terms)
                        invocation.return_value(
                            GLib.Variant('(as)', (results,)))

                    elif method_name == "GetResultMetas":
                        identifiers = parameters.unpack()[0]
                        metas = self.GetResultMetas(identifiers)
                        invocation.return_value(
                            GLib.Variant('(aa{sv})', (metas,)))

                    elif method_name == "ActivateResult":
                        identifier, terms, timestamp = parameters.unpack()
                        self.ActivateResult(identifier, terms, timestamp)
                        invocation.return_value(None)

                    elif method_name == "LaunchSearch":
                        terms, timestamp = parameters.unpack()
                        self.LaunchSearch(terms, timestamp)
                        invocation.return_value(None)

                    else:
                        invocation.return_error_literal(
                            Gio.dbus_error_quark(),
                            Gio.DBusError.UNKNOWN_METHOD,
                            f"Unknown method: {method_name}"
                        )

                except Exception as e:
                    print(f"ERROR in {method_name}: {e}")
                    invocation.return_error_literal(
                        Gio.dbus_error_quark(),
                        Gio.DBusError.FAILED,
                        str(e)
                    )

            # Register the object
            self._registration_id = self._bus.register_object(
                "/me/iepure/devtoolbox/SearchProvider",
                interface_info,
                handle_method_call,
                None,  # get_property
                None   # set_property
            )

            print(f"DEBUG: Registered object with ID: {self._registration_id}")

            # Own the name
            name_id = Gio.bus_own_name(
                Gio.BusType.SESSION,
                "me.iepure.devtoolbox.SearchProvider",
                Gio.BusNameOwnerFlags.NONE,
                None,  # bus_acquired_closure
                None,  # name_acquired_closure
                None   # name_lost_closure
            )

            print(f"DEBUG: Name ownership ID: {name_id}")
            print("DEBUG: Search provider service started. Waiting for requests...")

            self._reset_idle_timer()

            # Run the main loop
            self._main_loop = GLib.MainLoop()
            self._main_loop.run()

        except Exception as e:
            print(f"ERROR: Failed to start search provider: {e}")
            return 1

        return 0

File: src/core/search_provider.rs
/*
 * search_provider.rs
 *
 * Copyright (C) 2026 Alessandro Iepure
 * SPDX-License-Identifier: GPL-3.0-or-later
*/

use gtk::{
    gio::{self, prelude::*, subclass::prelude::*, DBusError},
    glib,
};

use std::{cell::RefCell, collections::HashMap, time::Instant};

// Constants
const IDLE_TIMEOUT_SECS: u32 = 10;
const CACHE_DURATION_SECS: u64 = 300;

const SEARCH_PROVIDER_INTERFACE: &str = "org.gnome.Shell.SearchProvider2";
const SEARCH_PROVIDER_OBJECT_PATH: &str = "/me/iepure/devtoolbox/SearchProvider";

const SEARCH_PROVIDER_XML: &str =
    include_str!("../../data/me.iepure.devtoolbox.SearchProvider.xml");

// Variant types for D-Bus method arguments
#[derive(Debug, glib::Variant)]
struct GetInitialResultSetCall {
    terms: Vec<String>,
}

#[derive(Debug, glib::Variant)]
struct GetSubsearchResultSetCall {
    previous_results: Vec<String>,
    terms: Vec<String>,
}

#[derive(Debug, glib::Variant)]
struct GetResultMetasCall {
    identifiers: Vec<String>,
}

#[derive(Debug, glib::Variant)]
struct ActivateResultCall {
    identifier: String,
    terms: Vec<String>,
    timestamp: u32,
}

#[derive(Debug, glib::Variant)]
struct LaunchSearchCall {
    terms: Vec<String>,
    timestamp: u32,
}

// Method call enum
#[derive(Debug)]
enum SearchProviderMethod {
    GetInitialResultSet(GetInitialResultSetCall),
    GetSubsearchResultSet(GetSubsearchResultSetCall),
    GetResultMetas(GetResultMetasCall),
    ActivateResult(ActivateResultCall),
    LaunchSearch(LaunchSearchCall),
}

impl DBusMethodCall for SearchProviderMethod {
    fn parse_call(
        _obj_path: &str,
        _interface: Option<&str>,
        method: &str,
        params: glib::Variant,
    ) -> Result<Self, glib::Error> {
        match method {
            "GetInitialResultSet" => params
                .get::<GetInitialResultSetCall>()
                .map(Self::GetInitialResultSet)
                .ok_or_else(|| glib::Error::new(DBusError::InvalidArgs, "Invalid parameters")),

            "GetSubsearchResultSet" => params
                .get::<GetSubsearchResultSetCall>()
                .map(Self::GetSubsearchResultSet)
                .ok_or_else(|| glib::Error::new(DBusError::InvalidArgs, "Invalid parameters")),

            "GetResultMetas" => params
                .get::<GetResultMetasCall>()
                .map(Self::GetResultMetas)
                .ok_or_else(|| glib::Error::new(DBusError::InvalidArgs, "Invalid parameters")),

            "ActivateResult" => params
                .get::<ActivateResultCall>()
                .map(Self::ActivateResult)
                .ok_or_else(|| glib::Error::new(DBusError::InvalidArgs, "Invalid parameters")),

            "LaunchSearch" => params
                .get::<LaunchSearchCall>()
                .map(Self::LaunchSearch)
                .ok_or_else(|| glib::Error::new(DBusError::InvalidArgs, "Invalid parameters")),

            _ => Err(glib::Error::new(
                DBusError::UnknownMethod,
                &format!("Unknown method: {}", method),
            )),
        }
    }
}

// Internal state
mod imp {

    use crate::tools;

    use super::*;

    pub struct SearchProviderApp {
        registration_id: RefCell<Option<gio::RegistrationId>>,
        idle_timeout_id: RefCell<Option<glib::SourceId>>,
        search_cache: RefCell<HashMap<String, (Vec<String>, Instant)>>,
    }

    impl Default for SearchProviderApp {
        fn default() -> Self {
            Self {
                registration_id: RefCell::new(None),
                idle_timeout_id: RefCell::new(None),
                search_cache: RefCell::new(HashMap::new()),
            }
        }
    }

    #[glib::object_subclass]
    impl ObjectSubclass for SearchProviderApp {
        const NAME: &'static str = "SearchProviderApp";
        type Type = super::SearchProviderApp;
        type ParentType = gio::Application;
    }

    impl ObjectImpl for SearchProviderApp {
        fn constructed(&self) {
            self.parent_constructed();
        }
    }

    impl ApplicationImpl for SearchProviderApp {
        fn dbus_register(
            &self,
            connection: &gio::DBusConnection,
            object_path: &str,
        ) -> Result<(), glib::Error> {
            self.parent_dbus_register(connection, object_path)?;

            // Parse DBus interface
            let node_info = gio::DBusNodeInfo::for_xml(SEARCH_PROVIDER_XML).map_err(|e| {
                glib::Error::new(DBusError::Failed, &format!("XML parse error: {e}"))
            })?;
            let interface = node_info
                .lookup_interface(SEARCH_PROVIDER_INTERFACE)
                .ok_or_else(|| {
                    glib::Error::new(
                        DBusError::Failed,
                        &format!("Interface {SEARCH_PROVIDER_INTERFACE} not found in XML"),
                    )
                })?;

            let app_down = self.obj().downgrade();

            let id = connection
                .register_object(SEARCH_PROVIDER_OBJECT_PATH, &interface)
                .typed_method_call::<SearchProviderMethod>()
                .invoke_and_return_future_local(move |_connection, _sender, call| {
                    let app = app_down.clone();
                    async move {
                        // Guard against the app being dropped
                        let Some(app) = app.upgrade() else {
                            return Ok(None);
                        };

                        // Reset idle timeout
                        app.imp().reset_idle_timer();

                        match call {
                            SearchProviderMethod::GetInitialResultSet(params) => {
                                let results = app.imp().search_with_cache(&params.terms);
                                Ok(Some((results,).to_variant()))
                            }
                            SearchProviderMethod::GetSubsearchResultSet(params) => {
                                let results = app.imp().search_with_cache(&params.terms);
                                Ok(Some((results,).to_variant()))
                            }
                            SearchProviderMethod::GetResultMetas(params) => {
                                let metas = app.imp().build_result_metas(&params.identifiers);
                                Ok(Some(metas))
                            }
                            SearchProviderMethod::ActivateResult(params) => {
                                launch_app("--tool", &params.identifier);
                                Ok(None)
                            }
                            SearchProviderMethod::LaunchSearch(params) => {
                                launch_app("--search", &params.terms.join(" "));
                                Ok(None)
                            }
                        }
                    }
                })
                .build()?;

            self.registration_id.replace(Some(id));
            Ok(())
        }

        fn dbus_unregister(&self, connection: &gio::DBusConnection, object_path: &str) {
            self.parent_dbus_unregister(connection, object_path);
            if let Some(id) = self.registration_id.take() {
                if connection.unregister_object(id).is_err() {
                    eprintln!("Failed to unregister D-Bus object at {}", object_path);
                }
            }
        }
    }

    impl SearchProviderApp {
        fn reset_idle_timer(&self) {
            // Clear existing timer
            if let Some(id) = self.idle_timeout_id.take() {
                id.remove();
            }

            let app = self.obj().downgrade();
            let id = glib::timeout_add_seconds_local(IDLE_TIMEOUT_SECS, move || {
                if let Some(app) = app.upgrade() {
                    app.quit();
                }
                glib::ControlFlow::Break
            });

            self.idle_timeout_id.replace(Some(id));
        }

        fn search_with_cache(&self, terms: &[String]) -> Vec<String> {
            let cache_key = terms.join("|");
            let now = Instant::now();

            // Check cache
            {
                let cache = self.search_cache.borrow();
                if let Some((results, timestamp)) = cache.get(&cache_key) {
                    if now.duration_since(*timestamp).as_secs() < CACHE_DURATION_SECS {
                        return results.clone();
                    }
                }
            }

            // Do the search
            let search_terms: Vec<&str> = terms.iter().map(|s| s.as_str()).collect();
            let results: Vec<String> = tools::search_tools(&search_terms)
                .into_iter()
                .map(|tool| tool.id.to_string())
                .collect();

            // Cache result
            self.search_cache
                .borrow_mut()
                .insert(cache_key, (results.clone(), now));

            // Delete stale entries
            self.search_cache.borrow_mut().retain(|_, (_, timestamp)| {
                now.duration_since(*timestamp).as_secs() < CACHE_DURATION_SECS
            });

            results
        }

        fn build_result_metas(&self, identifiers: &[String]) -> glib::Variant {
            let mut dicts: Vec<glib::Variant> = Vec::new();

            for tool_id in identifiers {
                let Some(meta) = tools::ALL_TOOLS.iter().find(|tool| tool.id == tool_id) else {
                    continue;
                };

                let dict = glib::VariantDict::new(None);
                dict.insert("id", &meta.id);
                dict.insert("name", &meta.title);
                dict.insert("description", &meta.description);

                // todo: figure out how to add icons to the search provider results

                dicts.push(dict.end());
            }

            let child_type = glib::VariantTy::new("a{sv}").expect("Built-in type a{sv}");
            glib::Variant::array_from_iter_with_type(&child_type, &dicts)
        }
    }
}

// Public wrapper
glib::wrapper! {
    pub struct SearchProviderApp(ObjectSubclass<imp::SearchProviderApp>)
        @extends gio::Application,
        @implements gio::ActionGroup, gio::ActionMap;
}

impl SearchProviderApp {
    pub fn new() -> Self {
        glib::Object::builder()
            .property("application-id", "me.iepure.devtoolbox.SearchProvider")
            .property("flags", gio::ApplicationFlags::HANDLES_COMMAND_LINE)
            .build()
    }
}

impl Default for SearchProviderApp {
    fn default() -> Self {
        Self::new()
    }
}

fn launch_app(flag: &str, value: &str) {
    let bin = std::env::args()
        .next()
        .unwrap_or_else(|| "devtoolbox".into());
    std::process::Command::new(&bin)
        .args([flag, value])
        .spawn()
        .expect("Application failed to start");
}
